### PR TITLE
Add install instructions for editable Python installation

### LINKED ISSUE #719: Editable install of Python interface broken
Using the following dockerfile
```docker
FROM ubuntu:22.04

ENV DEBIAN_FRONTEND=noninteractive
ARG TARGET_PLATFORM
RUN apt-get update && \
    apt-get install -y software-properties-common
RUN apt-get install -y git \
    cmake \
    gcc \
    g++ \
    ninja-build \
    libopenblas-dev \ 
    liblapack-dev

RUN apt-get install -y python3-pip
RUN python3 -m pip install -U setuptools pip
RUN python3 -m pip install git+https://github.com/FEniCS/ufl.git

ENV CMAKE_BUILD_PARALLEL_LEVEL 2
WORKDIR /src
RUN git clone https://github.com/FEniCS/basix.git
RUN cd basix && \
    cmake -G Ninja -B build-dir -DCMAKE_BUILD_TYPE=Debug -S cpp/ && \
    cmake --build build-dir && \ 
    cmake --install build-dir


RUN cd basix/python && \
    python3 -m pip install -e . -v


WORKDIR /tmp
RUN python3 -c "import basix"
```
results in
```bash
#15 [12/12] RUN python3 -c "import basix"
#15 0.369 <frozen importlib._bootstrap>:241: RuntimeWarning: nanobind: type 'LatticeType' was already registered!
#15 0.369 
#15 0.369 <frozen importlib._bootstrap>:241: RuntimeWarning: nanobind: type 'LatticeSimplexMethod' was already registered!
#15 0.369 
#15 0.369 <frozen importlib._bootstrap>:241: RuntimeWarning: nanobind: type 'PolynomialType' was already registered!
#15 0.369 
#15 0.369 <frozen importlib._bootstrap>:241: RuntimeWarning: nanobind: type 'MapType' was already registered!
#15 0.369 
#15 0.369 <frozen importlib._bootstrap>:241: RuntimeWarning: nanobind: type 'SobolevSpace' was already registered!
#15 0.369 
#15 0.369 <frozen importlib._bootstrap>:241: RuntimeWarning: nanobind: type 'QuadratureType' was already registered!
#15 0.369 
#15 0.370 <frozen importlib._bootstrap>:241: RuntimeWarning: nanobind: type 'CellType' was already registered!
#15 0.370 
#15 0.370 <frozen importlib._bootstrap>:241: RuntimeWarning: nanobind: type 'ElementFamily' was already registered!
#15 0.370 
#15 0.370 <frozen importlib._bootstrap>:241: RuntimeWarning: nanobind: type 'FiniteElement' was already registered!
#15 0.370 
#15 0.370 <frozen importlib._bootstrap>:241: RuntimeWarning: nanobind: type 'LagrangeVariant' was already registered!
#15 0.370 
#15 0.370 <frozen importlib._bootstrap>:241: RuntimeWarning: nanobind: type 'DPCVariant' was already registered!
#15 0.370 
#15 0.370 <frozen importlib._bootstrap>:241: RuntimeWarning: nanobind: type 'PolysetType' was already registered!
#15 0.370 
#15 0.370 Traceback (most recent call last):
#15 0.370   File "<string>", line 1, in <module>
#15 0.370   File "/src/basix/python/basix/__init__.py", line 7, in <module>
#15 0.370     from . import cell, finite_element, lattice, polynomials, quadrature, sobolev_spaces, variants
#15 0.370   File "/src/basix/python/basix/finite_element.py", line 3, in <module>
#15 0.370     from ._basixcpp import ElementFamily as _EF
#15 0.370 ImportError: cannot import name 'ElementFamily' from 'basix.finite_element._basixcpp' (/usr/local/lib/python3.10/dist-packages/basix/_basixcpp.cpython-310-x86_64-linux-gnu.so)
#15 0.375 free(): invalid pointer
#15 0.433 Aborted (core dumped)
#15 ERROR: process "/bin/sh -c python3 -c \"import basix\"" did not complete successfully: exit code: 134
------
 > [12/12] RUN python3 -c "import basix":
0.370 
0.370 Traceback (most recent call last):
0.370   File "<string>", line 1, in <module>
0.370   File "/src/basix/python/basix/__init__.py", line 7, in <module>
0.370     from . import cell, finite_element, lattice, polynomials, quadrature, sobolev_spaces, variants
0.370   File "/src/basix/python/basix/finite_element.py", line 3, in <module>
0.370     from ._basixcpp import ElementFamily as _EF
0.370 ImportError: cannot import name 'ElementFamily' from 'basix.finite_element._basixcpp' (/usr/local/lib/python3.10/dist-packages/basix/_basixcpp.cpython-310-x86_64-linux-gnu.so)
0.375 free(): invalid pointer
0.433 Aborted (core dumped)
------
Dockerfile:35
--------------------
  33 |     
  34 |     WORKDIR /tmp
  35 | >>> RUN python3 -c "import basix"
  36 |     
  37 |     # libgmp-dev
--------------------
ERROR: failed to solve: process "/bin/sh -c python3 -c \"import basix\"" did not complete successfully: exit code: 134

```
full log attached
[log.txt](https://github.com/FEniCS/basix/files/13185853/log.txt)

### PR DESCRIPTION
**Pull request by jorgensd**
Tested with:
```python
FROM ubuntu:22.04

ENV DEBIAN_FRONTEND=noninteractive
ARG TARGET_PLATFORM
RUN apt-get update && \
    apt-get install -y software-properties-common
RUN apt-get install -y git \
    cmake \
    gcc \
    g++ \
    ninja-build \
    libopenblas-dev \ 
    liblapack-dev
ENV DEB_PYTHON_INSTALL_LAYOUT=deb_system
RUN apt-get install -y python3-pip
RUN python3 -m pip install -U setuptools pip
RUN python3 -m pip install git+https://github.com/FEniCS/ufl.git

ENV CMAKE_BUILD_PARALLEL_LEVEL 2
WORKDIR /src
RUN git clone --branch=dokken/basix-install-instructions --single-branch https://github.com/FEniCS/basix.git
WORKDIR /src/basix
RUN cmake -G Ninja -B build-dir -DCMAKE_BUILD_TYPE=Debug -S cpp/ && \
    cmake --build build-dir && \ 
    cmake --install build-dir

WORKDIR /src/basix/python
RUN python3 -m pip -v install -r build-requirements.txt
RUN python3 -m pip -v install --config-settings=build-dir="build" --config-settings=cmake.build-type="Debug" --no-build-isolation -e .

WORKDIR /tmp
RUN python3 -c "import basix"
```
and
```python
FROM ubuntu:22.04

ENV DEBIAN_FRONTEND=noninteractive
ARG TARGET_PLATFORM
RUN apt-get update && \
    apt-get install -y software-properties-common
RUN apt-get install -y git \
    cmake \
    gcc \
    g++ \
    ninja-build \
    libopenblas-dev \ 
    liblapack-dev
ENV DEB_PYTHON_INSTALL_LAYOUT=deb_system
RUN apt-get install -y python3-pip
RUN python3 -m pip install -U setuptools pip
RUN python3 -m pip install git+https://github.com/FEniCS/ufl.git

ENV CMAKE_BUILD_PARALLEL_LEVEL 2
WORKDIR /src
RUN git clone --branch=dokken/basix-install-instructions --single-branch https://github.com/FEniCS/basix.git
WORKDIR /src/basix
RUN cmake -G Ninja -B build-dir -DCMAKE_BUILD_TYPE=Debug -S cpp/ && \
    cmake --build build-dir && \ 
    cmake --install build-dir

WORKDIR /src/basix
RUN python3 -m pip -v install -r build-requirements.txt
RUN python3 -m pip -v install --config-settings=build-dir="build" --config-settings=cmake.build-type="Debug" --no-build-isolation -e .

WORKDIR /tmp
RUN python3 -c "import basix"
```
Resolves #719

### CODE REVIEW

File: python/README.rst [hunk 1]
```diff
@@ -0,0 +1,20 @@
+Basix Python interface
+========================
+
+This document explains how to install the Basix Python interface. Note that
```
Comment (jhale) [inline python/README.rst:4]: I would add a note along the lines that regular users should use the top-level install.

File: python/README.rst [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+this interface must be built without PEP517 build isolation by passing
+`--no-build-isolation` flag to `pip install`.
```
Comment (jorgensd) [inline python/README.rst:6]:
```suggestion
This document explains how to install the Basix Python interface, given that the C++ interface has been built.
This is meant for development purposes, and regular users should use the top-level install instructions.
```
What do you think @jhale ?

**Review by garth-wells — COMMENTED** [review]
See earlier PR - #720.

**Comment by jhale** [comment]
I think this PR is a little more complete than #720

**Comment by francesco-ballarin** [comment]
As user, I struggle to understand`*` why build isolation is required with editable installs, but not with standard one. Maybe the reason for this is worth mentioning in the README?

`*` I do understand https://github.com/FEniCS/dolfinx/pull/2707 instead, in the sense that with build isolation even the standard install would fail to locate `mpi4py` and `petsc4py`.

**Comment by garth-wells** [comment]
> I think this PR is a little more complete than #720

It has too much text. Too much to read and too much that can go out of date. The instructions are for developers, who can do a bit of work themselves.

**Comment by jorgensd** [comment]
> > I think this PR is a little more complete than #720
> 
> It has too much text. Too much to read and too much that can go out of date. The instructions are for developers, who can do a bit of work themselves.

I dont this this is fair. It is far from trivial to get this to work. The other PR doesn’t even have valid instructions, as it is missing pins and build dependencies

**Comment by jorgensd** [comment]
> As user, I struggle to understand`*` why build isolation is required with editable installs, but not with standard ones. Maybe the reason for this is worth mentioning in the README?
> 
> `*` I do understand [FEniCS/dolfinx#2707](https://github.com/FEniCS/dolfinx/pull/2707) instead, in the sense that with build isolation even the standard install would fail to locate `mpi4py` and `petsc4py`.

I'm not sure what the cause of the issue is, but if you don't use build-isolation, you get the following error when import basix:
``bash
 > [14/14] RUN python3 -c "import basix":                                                                                                                                                                      
0.408 <frozen importlib._bootstrap>:241: RuntimeWarning: nanobind: type 'LatticeType' was already registered!                                                                                                  
0.408                                                                                                                                                                                                          
0.409 <frozen importlib._bootstrap>:241: RuntimeWarning: nanobind: type 'LatticeSimplexMethod' was already registered!
0.409 
0.410 <frozen importlib._bootstrap>:241: RuntimeWarning: nanobind: type 'PolynomialType' was already registered!
0.410 
0.410 <frozen importlib._bootstrap>:241: RuntimeWarning: nanobind: type 'MapType' was already registered!
0.410 
0.410 <frozen importlib._bootstrap>:241: RuntimeWarning: nanobind: type 'SobolevSpace' was already registered!
0.410 
0.410 <frozen importlib._bootstrap>:241: RuntimeWarning: nanobind: type 'QuadratureType' was already registered!
0.410 
0.411 <frozen importlib._bootstrap>:241: RuntimeWarning: nanobind: type 'CellType' was already registered!
0.411 
0.411 <frozen importlib._bootstrap>:241: RuntimeWarning: nanobind: type 'ElementFamily' was already registered!
0.411 
0.411 <frozen importlib._bootstrap>:241: RuntimeWarning: nanobind: type 'FiniteElement' was already registered!
0.411 
0.412 <frozen importlib._bootstrap>:241: RuntimeWarning: nanobind: type 'LagrangeVariant' was already registered!
0.412 
0.412 <frozen importlib._bootstrap>:241: RuntimeWarning: nanobind: type 'DPCVariant' was already registered!
0.412 
0.412 <frozen importlib._bootstrap>:241: RuntimeWarning: nanobind: type 'PolysetType' was already registered!
0.412 
0.413 Traceback (most recent call last):
0.413   File "<string>", line 1, in <module>
0.414   File "/src/basix/python/basix/__init__.py", line 7, in <module>
0.414     from . import cell, finite_element, lattice, polynomials, quadrature, sobolev_spaces, variants
0.414   File "/src/basix/python/basix/finite_element.py", line 3, in <module>
0.415     from ._basixcpp import ElementFamily as _EF
0.415 ImportError: cannot import name 'ElementFamily' from 'basix.finite_element._basixcpp' (/usr/lib/python3/dist-packages/basix/_basixcpp.cpython-310-x86_64-linux-gnu.so)
0.419 free(): invalid pointer
0.507 Aborted (core dumped)
------
```

**Comment by francesco-ballarin** [comment]
> I'm not sure what the cause of the issue is, but if you don't use build-isolation, you get the following error when import basix:

I guess that my question is more: how come that `pip install .` (without `-e`) works correctly?

**Comment by jhale** [comment]
@francesco-ballarin It is a bug in scikit-build-core latest version for editable builds that will be fixed upstream at some point.

**Comment by jorgensd** [comment]
> @francesco-ballarin It is a bug in scikit-build-core latest version for editable builds that will be fixed upstream at some point.

Note that even pining scikit-build-core doesn’t help wrt removing build isolation. So there are two scikit bugs here:
1. Need build isolation for editable install 
2. https://github.com/scikit-build/scikit-build-core/issues/532

**Comment by jorgensd** [comment]
A compromise would be to use split install in one of the CIs, as it would then be easily maintainable and easily accessed by developers.

**Comment by mscroggs** [comment]
> split install

I think that [this run](https://github.com/FEniCS/basix/blob/3b4cf17b574c0f181f539daaa8b4be099ccae0c7/.github/workflows/pythonapp.yml#L143-L183) on the CI is trying to do this, but perhaps needs updating? Or do you mean something else by split install?

**Comment by jorgensd** [comment]
> > split install
> 
> I think that [this run](https://github.com/FEniCS/basix/blob/3b4cf17b574c0f181f539daaa8b4be099ccae0c7/.github/workflows/pythonapp.yml#L143-L183) on the CI is trying to do this, but perhaps needs updating? Or do you mean something else by split install?

It does a split install, but not a split editable installation. Most developers would probably use the editable install option (which has the requirements described above).

**Comment by mscroggs** [comment]
> > > split install
> > 
> > 
> > I think that [this run](https://github.com/FEniCS/basix/blob/3b4cf17b574c0f181f539daaa8b4be099ccae0c7/.github/workflows/pythonapp.yml#L143-L183) on the CI is trying to do this, but perhaps needs updating? Or do you mean something else by split install?
> 
> It does a split install, but not a split editable installation. Most developers would probably use the editable install option (which has the requirements described above).

We change it to editable. Or include both (is there any benefit to having both?)

**Comment by jorgensd** [comment]
> > > > split install
> > > 
> > > 
> > > I think that [this run](https://github.com/FEniCS/basix/blob/3b4cf17b574c0f181f539daaa8b4be099ccae0c7/.github/workflows/pythonapp.yml#L143-L183) on the CI is trying to do this, but perhaps needs updating? Or do you mean something else by split install?
> > 
> > 
> > It does a split install, but not a split editable installation. Most developers would probably use the editable install option (which has the requirements described above).
> 
> We change it to editable. Or include both (is there any benefit to having both?)

Well, it would give better coverage. In any case, we would need the build_requirements.txt to be Added to Basix.

File: python/README.rst [hunk 3]
```diff
@@ -0,0 +1,19 @@
+Basix Python interface
+========================
+
+This document explains how to install the Basix Python interface, given that the C++ interface has been built.
```
Comment (garth-wells) [inline python/README.rst:4]: Shorten text.

File: python/build-requirements.txt
Comment (garth-wells) [inline python/build-requirements.txt:1]: See earlier version comments.

File: python/README.rst [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+This is meant for development purposes, and regular users should use the top-level install instructions.
+
+1. Build and install the Basix C++ library in the usual way.
+
+2. Ensure the build time requirements are installed::
+
+     python3 -m pip -v install -r build-requirements.txt
```
Comment (garth-wells) [inline python/README.rst:11]: Use `pip` (shorter) and no need for `-v`.

File: python/README.rst [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+3. Build Basix Python interface::
+
+    python3 -m pip -v install --no-build-isolation .
```
Comment (garth-wells) [inline python/README.rst:15]: See above comments.

**Review by garth-wells — CHANGES_REQUESTED** [review]

**Comment by garth-wells** [comment]
Closing this as the editable issue was fixed by #727. Can be re-opened with any refinements of the docs.

**Pull request closed without merging**